server/src-lib/Hasura/GraphQL/Execute.hs

```diff
@@ -486,76 +466,57 @@ insertBatchResults remoteValue batch hasuraResp =
-                 show (OJ.toEncJSON (OJ.Object hasuraHash)))
-            (theRemoteHash:remainingRemoteHashes) ->
-              Right
-                ( (foldl'
-                     (flip OJ.delete)
-                     (OJ.insert
-                        ( batchRelationshipKeyIndex batch
-                        , batchRelationshipKeyToMake batch)
-                        (peelOffNestedFields
-                           (batchNestedFields batch)
-                           (snd (theRemoteHash)))
-                        hasuraHash)
-                     (batchPhantoms batch))
-                , remainingRemoteHashes)
-    inHashmap (RelFieldPath ((idx, G.Alias (G.Name key)) Seq.:<| rest)) hasuraHash remoteHash =
-      case OJ.lookup key hasuraHash of
-        Nothing ->
-          Left
-            ("Couldn't find expected key " <> show key <> " in " <>
-             show (OJ.toEncJSON (OJ.Object hasuraHash)) <>
-             ", while traversing " <>
-             show (OJ.toEncJSON (OJ.Object hasuraHash0)))
-        Just currentValue ->
-          fmap
-            (\(newValue, remoteCandidates) ->
-               (OJ.insert (idx, key) newValue hasuraHash, remoteCandidates))
-            (inValue rest currentValue remoteHash)
+    inHashmap (RelFieldPath p) hasuraHash remoteHash = case p of
+      Seq.Empty ->
+        case remoteHash of
+          [] -> Left $ err <> show (OJ.toEncJSON $ OJ.Object hasuraHash)
+            where err = case cardinality of 
+                          One -> "Expected one remote object but got none, while traversing " 
+                          Many -> "Expected many objects but got none, while traversing " 
+
+          (theRemoteHash:remainingRemoteHashes) 
+            | cardinality == One && not (null remainingRemoteHashes) ->
+                Left $
+                  "Expected one remote object but got many, while traversing " <> showHash hasuraHash
+            | otherwise -> 
+                Right
+                  ( (foldl'
+                       (flip OJ.delete)
+                       (OJ.insert
+                          (batchRelationshipKeyIndex, batchRelationshipKeyToMake)
+                          (peelOffNestedFields batchNestedFields (snd theRemoteHash))
+                          hasuraHash)
+                       batchPhantoms)
+                  , remainingRemoteHashes)
+ 
+      ((idx, G.Alias (G.Name key)) Seq.:<| rest) ->
+        case OJ.lookup key hasuraHash of
+          Nothing ->
+            Left $ 
+              "Couldn't find expected key " <> show key <> " in " <> showHash hasuraHash <>
+              ", while traversing " <> showHash hasuraHash0
+          Just currentValue ->
+            first (\newValue -> OJ.insert (idx, key) newValue hasuraHash)
+              <$> inValue rest currentValue remoteHash
+      where
+        showHash = show . OJ.toEncJSON . OJ.Object
+
+    -- TODO what is this supposed to do, and what does the return values mean?
```

Comment:
`inValue` basically takes an aeson `Value` i.e. hasura result as value and puts the join data inside it.

The return value `Either String (OJ.Value, [(Text, OJ.Value)])` means the following:
```
\case 

Left errorString -> some error in joining remote data with hasura data
Right (respSoFar, remainingRemoteRows) -> when joining, we go to the end of the `RelFieldPath` of the hasura value and add a remote row. 

```

I am not sure how to prove the correctness of this algorithm. We should discuss.